In [29]:
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.metrics import accuracy_score
from sklearn.utils import resample
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

In [30]:
train_df = pd.read_csv(Path('Resources/2019loans.csv'))
test_df = pd.read_csv(Path('Resources/2020Q1loans.csv'))

In [31]:
# Quick Look at train_df
train_df.head()

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0     7000.0    0.1894       256.38       MORTGAGE     75000.0   
1    40000.0    0.1614       975.71       MORTGAGE    102000.0   
2    11000.0    0.2055       294.81           RENT     45000.0   
3     4000.0    0.1612       140.87       MORTGAGE     38000.0   
4    14000.0    0.1797       505.93       MORTGAGE     43000.0   

  verification_status pymnt_plan    dti  delinq_2yrs  inq_last_6mths  ...  \
0        Not Verified          n  28.62          0.0             2.0  ...   
1     Source Verified          n  11.72          2.0             0.0  ...   
2            Verified          n  37.25          1.0             3.0  ...   
3        Not Verified          n  42.89          1.0             0.0  ...   
4     Source Verified          n  22.16          1.0             0.0  ...   

   percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  \
0              87.5                   0.0        0.0         352260.0   
1               0.0                   0.0        0.0         294664.0   
2               7.7                   0.0        0.0          92228.0   
3             100.0                   0.0        0.0         284273.0   
4              25.0                   0.0        0.0         120280.0   

  total_bal_ex_mort  total_bc_limit  total_il_high_credit_limit  \
0           62666.0         35000.0                     10000.0   
1          109911.0          9000.0                     71044.0   
2           36007.0         33000.0                     46328.0   
3           52236.0         13500.0                     52017.0   
4           88147.0         33300.0                     78680.0   

   hardship_flag  debt_settlement_flag    target  
0              N                     N  low_risk  
1              N                     N  low_risk  
2              N                     N  low_risk  
3              N                     N  low_risk  
4              N                     N  low_risk  

[5 rows x 84 columns]

In [32]:
# Quick Look at test_df
test_df.head()

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0    40000.0    0.1033       856.40           RENT    128700.0   
1    24450.0    0.1430       572.72       MORTGAGE     44574.0   
2    13500.0    0.1430       316.23            OWN     60000.0   
3    10625.0    0.1774       268.31           RENT     60000.0   
4     6375.0    0.1862       232.46           RENT     60000.0   

  verification_status pymnt_plan    dti  delinq_2yrs  inq_last_6mths  ...  \
0     Source Verified          n  12.47          0.0             1.0  ...   
1        Not Verified          n  15.05          0.0             1.0  ...   
2        Not Verified          n  28.72          0.0             0.0  ...   
3            Verified          n  15.70          0.0             4.0  ...   
4     Source Verified          n  35.50          0.0             0.0  ...   

   percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  \
0              57.1                   0.0        0.0          63915.0   
1               0.0                   0.0        0.0         136425.0   
2               0.0                   0.0        0.0          82124.0   
3              20.0                   0.0        0.0          54855.0   
4              75.0                   0.0        0.0          90445.0   

  total_bal_ex_mort  total_bc_limit  total_il_high_credit_limit  \
0           49510.0         49400.0                     14515.0   
1           19439.0         15500.0                     18925.0   
2           65000.0          5400.0                     61724.0   
3           50335.0         23200.0                     26255.0   
4           56541.0         15300.0                     72345.0   

   hardship_flag  debt_settlement_flag    target  
0              Y                     N  low_risk  
1              N                     N  low_risk  
2              Y                     N  low_risk  
3              N                     N  low_risk  
4              N                     N  low_risk  

[5 rows x 84 columns]

In [34]:
# Convert categorical data to numeric and separate target feature for training data
# (use 2019 loans data and pd.get_dummies() to convert the categorical data to numeric columns)
X = train_df.drop('target', axis=1)
y_train = LabelEncoder().fit_transform(train_df['target'])
X_train = pd.get_dummies(X)
print(X_train.columns)
X_train

Index(['loan_amnt', 'int_rate', 'installment', 'annual_inc', 'dti',
       'delinq_2yrs', 'inq_last_6mths', 'open_acc', 'pub_rec', 'revol_bal',
       'total_acc', 'out_prncp', 'out_prncp_inv', 'total_pymnt',
       'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
       'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
       'last_pymnt_amnt', 'collections_12_mths_ex_med', 'policy_code',
       'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m',
       'open_act_il', 'open_il_12m', 'open_il_24m', 'mths_since_rcnt_il',
       'total_bal_il', 'il_util', 'open_rv_12m', 'open_rv_24m', 'max_bal_bc',
       'all_util', 'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_12m',
       'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util',
       'chargeoff_within_12_mths', 'delinq_amnt', 'mo_sin_old_il_acct',
       'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl',
       'mort_acc', 'mths_since_recent_bc', 'mths_since_recent_in

loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0         7000.0    0.1894       256.38     75000.0  28.62          0.0   
1        40000.0    0.1614       975.71    102000.0  11.72          2.0   
2        11000.0    0.2055       294.81     45000.0  37.25          1.0   
3         4000.0    0.1612       140.87     38000.0  42.89          1.0   
4        14000.0    0.1797       505.93     43000.0  22.16          1.0   
...          ...       ...          ...         ...    ...          ...   
12175    19975.0    0.2565       801.09     28000.0  28.42          0.0   
12176    15000.0    0.1774       540.34     50000.0  23.43          4.0   
12177     3600.0    0.1862       131.28     60000.0  28.80          0.0   
12178    15000.0    0.0881       475.68     62000.0  11.44          0.0   
12179    15000.0    0.1774       540.34     54080.0  21.66          1.0   

       inq_last_6mths  open_acc  pub_rec  revol_bal  ...  \
0                 2.0      20.0      0.0    40414.0  ...   
1                 0.0      10.0      0.0    43531.0  ...   
2                 3.0      23.0      0.0     8242.0  ...   
3                 0.0       7.0      0.0    12767.0  ...   
4                 0.0      22.0      0.0    11182.0  ...   
...               ...       ...      ...        ...  ...   
12175             0.0      15.0      0.0     9679.0  ...   
12176             0.0      16.0      0.0     3193.0  ...   
12177             1.0      14.0      0.0    11804.0  ...   
12178             0.0       5.0      0.0      157.0  ...   
12179             1.0       5.0      0.0     8349.0  ...   

       verification_status_Verified  pymnt_plan_n  initial_list_status_f  \
0                                 0             1                      1   
1                                 0             1                      0   
2                                 1             1                      0   
3                                 0             1                      0   
4                                 0             1                      0   
...                             ...           ...                    ...   
12175                             0             1                      0   
12176                             1             1                      0   
12177                             0             1                      0   
12178                             0             1                      0   
12179                             0             1                      1   

       initial_list_status_w  application_type_Individual  \
0                          0                            1   
1                          1                            1   
2                          1                            1   
3                          1                            0   
4                          1                            1   
...                      ...                          ...   
12175                      1                            1   
12176                      1                            1   
12177                      1                            1   
12178                      1                            0   
12179                      0                            1   

       application_type_Joint App  hardship_flag_N  hardship_flag_Y  \
0                               0                1                0   
1                               0                1                0   
2                               0                1                0   
3                               1                1                0   
4                               0                1                0   
...                           ...              ...              ...   
12175                           0                1                0   
12176                           0                1                0   
12177                           0                1                0   
12178                        

In [35]:
# Convert categorical data to numeric and separate target feature for testing data
# (use 2020 loans data and pd.get_dummies() to convert the categorical data to numeric columns)
X = test_df.drop('target', axis=1)
y_test = LabelEncoder().fit_transform(test_df['target'])
X_test = pd.get_dummies(X)
print(X_test.columns)
X_test

Index(['loan_amnt', 'int_rate', 'installment', 'annual_inc', 'dti',
       'delinq_2yrs', 'inq_last_6mths', 'open_acc', 'pub_rec', 'revol_bal',
       'total_acc', 'out_prncp', 'out_prncp_inv', 'total_pymnt',
       'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
       'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
       'last_pymnt_amnt', 'collections_12_mths_ex_med', 'policy_code',
       'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m',
       'open_act_il', 'open_il_12m', 'open_il_24m', 'mths_since_rcnt_il',
       'total_bal_il', 'il_util', 'open_rv_12m', 'open_rv_24m', 'max_bal_bc',
       'all_util', 'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_12m',
       'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util',
       'chargeoff_within_12_mths', 'delinq_amnt', 'mo_sin_old_il_acct',
       'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl',
       'mort_acc', 'mths_since_recent_bc', 'mths_since_recent_in

loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0       40000.0    0.1033       856.40    128700.0  12.47          0.0   
1       24450.0    0.1430       572.72     44574.0  15.05          0.0   
2       13500.0    0.1430       316.23     60000.0  28.72          0.0   
3       10625.0    0.1774       268.31     60000.0  15.70          0.0   
4        6375.0    0.1862       232.46     60000.0  35.50          0.0   
...         ...       ...          ...         ...    ...          ...   
4697    30000.0    0.1240       673.42    140480.0  15.74          0.0   
4698    24000.0    0.0756       747.22     50000.0  26.81          0.0   
4699    10000.0    0.2305       387.36     33000.0  38.51          0.0   
4700     8000.0    0.1862       205.86     38000.0  16.36          0.0   
4701    30000.0    0.2055      1123.34    180000.0  12.06          0.0   

      inq_last_6mths  open_acc  pub_rec  revol_bal  ...  \
0                1.0       8.0      0.0    38113.0  ...   
1                1.0       6.0      0.0     1665.0  ...   
2                0.0       8.0      0.0    13857.0  ...   
3                4.0      17.0      0.0     6216.0  ...   
4                0.0      13.0      0.0    12681.0  ...   
...              ...       ...      ...        ...  ...   
4697             0.0      20.0      0.0    23215.0  ...   
4698             0.0       9.0      0.0      459.0  ...   
4699             2.0       7.0      0.0     6342.0  ...   
4700             1.0       8.0      1.0    11636.0  ...   
4701             0.0       8.0      0.0     4771.0  ...   

      verification_status_Source Verified  verification_status_Verified  \
0                                       1                             0   
1                                       0                             0   
2                                       0                             0   
3                                       0                             1   
4                                       1                             0   
...                                   ...                           ...   
4697                                    1                             0   
4698                                    0                             0   
4699                                    0                             1   
4700                                    1                             0   
4701                                    1                             0   

      pymnt_plan_n  initial_list_status_f  initial_list_status_w  \
0                1                      0                      1   
1                1                      0                      1   
2                1                      0                      1   
3                1                      0                      1   
4                1                      0                      1   
...            ...                    ...                    ...   
4697             1                      1                      0   
4698             1                      0                      1   
4699             1                      1                      0   
4700             1                      0                      1   
4701             1                      1                      0   

      application_type_Individual  application_type_Joint App  \
0                               1                           0   
1                               1                           0   
2                               1                           0   
3                               1                           0   
4                               1                           0   
...                           ...                         ...   
4697                            1                           0   
4698                            1                           0   
4699                            1                           0   
4700                            1       

In [36]:
# add missing dummy variables to testing set
X_test['debt_settlement_flag_Y'] = 0
print(X_test.columns)
X_test

Index(['loan_amnt', 'int_rate', 'installment', 'annual_inc', 'dti',
       'delinq_2yrs', 'inq_last_6mths', 'open_acc', 'pub_rec', 'revol_bal',
       'total_acc', 'out_prncp', 'out_prncp_inv', 'total_pymnt',
       'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
       'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
       'last_pymnt_amnt', 'collections_12_mths_ex_med', 'policy_code',
       'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m',
       'open_act_il', 'open_il_12m', 'open_il_24m', 'mths_since_rcnt_il',
       'total_bal_il', 'il_util', 'open_rv_12m', 'open_rv_24m', 'max_bal_bc',
       'all_util', 'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_12m',
       'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util',
       'chargeoff_within_12_mths', 'delinq_amnt', 'mo_sin_old_il_acct',
       'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl',
       'mort_acc', 'mths_since_recent_bc', 'mths_since_recent_in

loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0       40000.0    0.1033       856.40    128700.0  12.47          0.0   
1       24450.0    0.1430       572.72     44574.0  15.05          0.0   
2       13500.0    0.1430       316.23     60000.0  28.72          0.0   
3       10625.0    0.1774       268.31     60000.0  15.70          0.0   
4        6375.0    0.1862       232.46     60000.0  35.50          0.0   
...         ...       ...          ...         ...    ...          ...   
4697    30000.0    0.1240       673.42    140480.0  15.74          0.0   
4698    24000.0    0.0756       747.22     50000.0  26.81          0.0   
4699    10000.0    0.2305       387.36     33000.0  38.51          0.0   
4700     8000.0    0.1862       205.86     38000.0  16.36          0.0   
4701    30000.0    0.2055      1123.34    180000.0  12.06          0.0   

      inq_last_6mths  open_acc  pub_rec  revol_bal  ...  \
0                1.0       8.0      0.0    38113.0  ...   
1                1.0       6.0      0.0     1665.0  ...   
2                0.0       8.0      0.0    13857.0  ...   
3                4.0      17.0      0.0     6216.0  ...   
4                0.0      13.0      0.0    12681.0  ...   
...              ...       ...      ...        ...  ...   
4697             0.0      20.0      0.0    23215.0  ...   
4698             0.0       9.0      0.0      459.0  ...   
4699             2.0       7.0      0.0     6342.0  ...   
4700             1.0       8.0      1.0    11636.0  ...   
4701             0.0       8.0      0.0     4771.0  ...   

      verification_status_Verified  pymnt_plan_n  initial_list_status_f  \
0                                0             1                      0   
1                                0             1                      0   
2                                0             1                      0   
3                                1             1                      0   
4                                0             1                      0   
...                            ...           ...                    ...   
4697                             0             1                      1   
4698                             0             1                      0   
4699                             1             1                      1   
4700                             0             1                      0   
4701                             0             1                      1   

      initial_list_status_w  application_type_Individual  \
0                         1                            1   
1                         1                            1   
2                         1                            1   
3                         1                            1   
4                         1                            1   
...                     ...                          ...   
4697                      0                            1   
4698                      1                            1   
4699                      0                            1   
4700                      1                            1   
4701                      0                            1   

      application_type_Joint App  hardship_flag_N  hardship_flag_Y  \
0                              0                0                1   
1                              0                1                0   
2                              0                0                1   
3                              0                1                0   
4                              0                1                0   
...                          ...              ...              ...   
4697                           0                1                0   
4698                           0                1                0   
4699                           0                1                0   
4700                           0                1                0   
4701            

# Create and compare two models: a Logistic Regression and a Random Forests Classifier

## Prediction:
The Random Forests Classifier will perform better than the Logistic Regression because of the large number of variables at play, many of which may be noise. As such, an average of predictions from multiple trees should be better than a single tree with everything in it.

In [37]:
# Train the Logistic Regression model on the unscaled data and print the model score
classifier = LogisticRegression()
classifier.fit(X_train, y_train)
print(f"Training Data Score: {classifier.score(X_train, y_train)}")
print(f"Testing Data Score: {classifier.score(X_test, y_test)}")

Training Data Score: 0.6522988505747126
Testing Data Score: 0.5108464483198639


C:\Users\josep\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [38]:
# Train a Random Forest Classifier model and print the model score
clfs = []
scores = []
for i in range(50):
    X_train_bootstrap, y_train_bootstrap = resample(X_train, y_train, random_state=i)
    clf = DecisionTreeClassifier(splitter='random', random_state=i+200).fit(X_train_bootstrap, y_train_bootstrap)
    clfs.append(clf)
    y_preds = [clf.predict(X_test) for clf in clfs]
    y_pred = pd.DataFrame(y_preds).median().round()
    score = score = accuracy_score(y_test, y_pred)
    scores.append(score)
clf = RandomForestClassifier(random_state=1, n_estimators=50).fit(X_train, y_train)
print(f"Training Data Score: {clf.score(X_train, y_train)}")
print(f"Testing Data Score: {clf.score(X_test, y_test)}")

Training Data Score: 1.0
Testing Data Score: 0.6205869842620162


## Results and how they compared to my prediction:
The Random Forest Classifier performed better than the Logistic Regression model, as predicted, but both models had lower testing data scores than I expected.
The Logistic Regression model's Training Data Score was: 0.6522988505747126, and Testing Data Score was: 0.5108464483198639.
The Random Forest Classifier's Training Data Score was: 1.0, and Testing Data Score was: 0.6205869842620162.

In [39]:
# Scale the data using StandardScaler
scaler = StandardScaler().fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

## Predictions on how the scaling will affect the accuracy of the models:
The scaled data will result in better scores with both models, hopefully making up for the low initial scores. Scaling ensures the magnitude of a few features don't bias the model during training, which is likely here as we have large numeric data compared to the smaller numbers used by the get_dummies method.

In [40]:
# Train the Logistic Regression model on the scaled data and print the model score
classifier = LogisticRegression()
classifier.fit(X_train_scaled, y_train)
print(f"Training Data Score: {classifier.score(X_train_scaled, y_train)}")
print(f"Testing Data Score: {classifier.score(X_test_scaled, y_test)}")

Training Data Score: 0.710919540229885
Testing Data Score: 0.7598894087622289


C:\Users\josep\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [41]:
# Train a Random Forest Classifier model on the scaled data and print the model score
clfs = []
scores = []
for i in range(50):
    X_train_scaled_bootstrap, y_train_bootstrap = resample(X_train_scaled, y_train, random_state=i)
    clf = DecisionTreeClassifier(splitter='random', random_state=i+200).fit(X_train_scaled_bootstrap, y_train_bootstrap)
    clfs.append(clf)
    y_preds = [clf.predict(X_test_scaled) for clf in clfs]
    y_pred = pd.DataFrame(y_preds).median().round()
    score = score = accuracy_score(y_test, y_pred)
    scores.append(score)
clf = RandomForestClassifier(random_state=1, n_estimators=50).fit(X_train_scaled, y_train)
print(f"Training Data Score: {clf.score(X_train_scaled, y_train)}")
print(f"Testing Data Score: {clf.score(X_test_scaled, y_test)}")

Training Data Score: 1.0
Testing Data Score: 0.623139089749043


## Results and how they compared to my prediction:
The scaled data improved both models, as predicted, but improved the Logistic Regression model much more than the Random Forest Classifier model. The Training Data Score for the Random Forest Classifier Model remained the same and the Testing Data Score was nearly the same (only slightly increased). The increase to the Testing Data Score for the Logistic Regression model was so much, however, that it improved to above the Testing Data Score of the Random Forest Classifier model, making the scaled Logistic Regression model superior when it comes to Testing Data.
The Logistic Regression model's Training Data Score was: 0.710919540229885, and Testing Data Score was: 0.7598894087622289 when scaled, compared to the Training Data Score of: 0.6522988505747126, and Testing Data Score of: 0.5108464483198639, unscaled.
The Random Forest Classifier's Training Data Score was: 1.0, and Testing Data Score was: 0.623139089749043 when scaled, compared to the Training Data Score of: 1.0, and Testing Data Score of: 0.6205869842620162, unscaled.